### ISSUE TITLE
onPostHandler and a (strange) persistant route context

### ISSUE THREAD

**Issue by Kevinlearynet** [hapijs/hapi#2522]
We're using an adapted version of the assets plugin for mtime based auto-versioning of our CSS/JS files. The plugin reads a set of CSS/JS assets from a global config file, and also allows us to define additional assets on a route by route basis within a controller. This gives us the flexibility to use CSS/JS thats specific to a route. The source for a controller and the plugin can be found at:

https://gist.github.com/Kevinlearynet/9653f9f42496db1cbc64

The functionality works, but we're experiencing a strange issue with assets included from within a controller. On initial load a CSS asset will be included properly in the `<head>` section of the document, but subsequent requests will append an additional `<link>` to the `<head>` each time the page is refreshed. We're not sure how the controller defined assets are persisting like this.

We're using a `catbox-redis` implementation as our default cache, so in-memory storage shouldn't be in use anywhere. We see no values in the redis server, but I have a feeling we may be missing something obvious with the way the `onPostHandler` lifecycle hook is working. Anything anyone can provide to point us in the right direction is greatly appreciated.

**Comment by devinivy** [comment]
I _think_ I understand what's going wrong, and I don't believe it to be hapi-related.
If `options` ([ref](https://gist.github.com/Kevinlearynet/9653f9f42496db1cbc64#file-plugin-js-L42)) are shared across many requests (which it is), then the list of `css` for an `environment` will grow on [this line](https://gist.github.com/Kevinlearynet/9653f9f42496db1cbc64#file-plugin-js-L42), as `globalAssets` refers to an object that's shared among requests.  Relatedly, `_.union` isn't preserving uniqueness because it's comparing objects by reference.

TL;DR: `options[ environment ].css` grows on [this line](https://gist.github.com/Kevinlearynet/9653f9f42496db1cbc64#file-plugin-js-L42) each request.  Try replacing,
```
var globalAssets = options[ environment ];
/* with */
var globalAssets = _.clone(options[ environment ]);
```
to ensure that the `options` object wont be modified.

**Comment by Kevinlearynet** [comment]
Thanks devinivy! This resolved the issue for us. 

Where/how is the `options` value stored in order to persists across uncached refreshes like this? Just to make sure I understand the underlying reason why this was happening.

**Comment by devinivy** [comment]
The `options` object persists by means of being in the scope of the plugin (passed as an argument).  When you assign `globalAssets` it does not copy the values inside the `options` object into `globalAssets`– instead `globalAssets` becomes synonymous with using `options[ environment ]` itself.
